ARROW-3726: [Rust] Add CSV reader with example

andygrove opened the pull request:
This adds a CSV reader and an example that accessed the loaded data through the use of downcasting arrays to specific types. The CSV reader supports all primitive types + string (`List<u8>`).

andygrove commented:
@paddyhoran @sunchao I would appreciate a review of this. The code in the example for reading strings is ugly and required calling `unsafe`. I don't know if I'm missing something or whether we need to improve the API here?

paddyhoran commented on rust/examples/read_csv.rs:
```diff
@@ -0,0 +1,79 @@
+    let lng = batch
+        .column(2)
+        .as_any()
+        .downcast_ref::<PrimitiveArray<f64>>()
```
Are we sure that the alignment requires are met if we simple downcast from any?  Alignments are checked through our existing constructors but would not be checked if we downcast, right?

andygrove replied:
I have no idea to be honest. I am definitely looking for guidance here on how to access these arrays after they are built.

paddyhoran replied:
Actually, I think it's fine as you are using our builders internally.  I'll take a closer look when I get a chance.

Thanks @andygrove

paddyhoran commented on rust/src/array.rs:
```diff
@@ -163,6 +168,12 @@ macro_rules! def_primitive_array {
                 unsafe { *(self.raw_values().offset(i as isize)) }
             }
 
+            pub fn value_slice(&self, offset: i64, len: i64) -> &[$native_ty] {
```
We should probably add a doc-comment for `value_slice`

paddyhoran commented on rust/src/csvreader.rs:
```diff
@@ -0,0 +1,243 @@
+        // read a batch of rows into memory
+        let mut rows: Vec<StringRecord> = Vec::with_capacity(self.batch_size);
+        for _ in 0..self.batch_size {
+            match self.record_iter.next() {
```
Should we expose `record_iter` to the user?  This would allow them to skip records at the beginning of a file for example before reading the rest of the file.

andygrove replied:
I will think about this. There is already the constructor param to indicate whether there is a header row or not that needs to be skipped. Maybe we could do this as a separate enhancement PR.

paddyhoran replied:
I have a use case where there is meta-data inserted after the header but before the data and I have no way to control this.  Perhaps this is a rather narrow use case.  In any case, happy to revisit this.

paddyhoran commented on rust/src/csvreader.rs:
```diff
@@ -0,0 +1,243 @@
+                                }
+                                _ => {
+                                    list_builder.append(false).unwrap();
+                                }
```
We can update this to use `BinaryArrayBuilder` once merged.

andygrove replied:
I pulled in 3787 and changed this to create `BinaryArray`

andygrove commented:
@paddyhoran I added documentation as requested. I also renamed CsvFile to CsvReader. I will be adding CsvWriter in separate PR once this one is merged (I already have working code).

sunchao commented on rust/src/csvreader.rs:
```diff
@@ -0,0 +1,263 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Can we have a separate mod for csv, i.e., `rust/src/csv` and `rust/src/csv/reader.rs`?

andygrove replied:
done

sunchao commented:
Thanks @andygrove . I'll take a look at this today too.

paddyhoran commented:
@andygrove the windows CI needs to be updated to run the new example also.

sunchao commented on rust/src/csvreader.rs:
```diff
@@ -0,0 +1,263 @@
+
+impl CsvReader {
+    /// Read the next batch of rows
+    pub fn next(&mut self) -> Option<Result<Arc<RecordBatch>, ArrowError>> {
```
Two points:
1. Can we return `RecordBatch` instead of `Arc<RecordBatch>`? we are creating unique reference for the record batches here and can let the caller to decide whether to make it a `Arc`.
2. We should use `error.Result` instead of `Result<T, ArrowError>`.

andygrove replied:
done

sunchao commented on rust/src/csvreader.rs:
```diff
@@ -0,0 +1,263 @@
+                    rows.push(r);
+                }
+                Some(Err(_)) => {
+                    return Some(Err(ArrowError::ParseError(
```
Perhaps we should surface this error into the `ParseError`.

sunchao commented on rust/src/csvreader.rs:
```diff
@@ -0,0 +1,263 @@
+        }
+
+        // return early if no data was loaded
+        if rows.len() == 0 {
```
nit: use `is_empty()` instead of `len() == 0`? it is more explicit.

andygrove replied:
done

sunchao commented on rust/src/csvreader.rs:
```diff
@@ -0,0 +1,263 @@
+                .collect(),
+        };
+
+        let arrays: Result<Vec<ArrayRef>, ArrowError> = projection
```
here too: we can use `error.Result`.

andygrove replied:
done

sunchao commented on rust/src/csvreader.rs:
```diff
@@ -0,0 +1,263 @@
+                    &DataType::UInt16 => build_primitive_array!(rows, i, u16),
+                    &DataType::UInt32 => build_primitive_array!(rows, i, u32),
+                    &DataType::UInt64 => build_primitive_array!(rows, i, u64),
+                    &DataType::Float16 => build_primitive_array!(rows, i, f32),
```
This doesn't look right - half float is stored with 2 bytes but here we are making them `f32` (4 bytes). Maybe we should leave it as a TODO for now.

andygrove replied:
I removed support for `Float16` so this would now fail with an unsupported data type error, which seems sensible

sunchao commented on rust/src/csv/mod.rs:
```diff
@@ -0,0 +1 @@
+pub mod reader;
```
License header in this file too. Also, I wonder if it's possible to rename `CsvReader` to `Reader` and export it on the module level, so people can import it with `use arrow::csv::Reader` instead of the current `use arrow::csv::reader::CsvReader`, which seems a little redundant.
